### PR TITLE
Fix 6534, handle unpack of assign-like tuples.

### LINKED ISSUE #6534: bytecode analysis failure with `BUILD_TUPLE_UNPACK`.
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

```python
from numba import njit

@njit
def foo(x):
    return (*x,)

foo((1,2))
```
does this:
```
<path>/numba/core/byteflow.py in _build_tuple_unpack(self, state, inst)
    846         state.append(inst, tuples=tuples, temps=temps)
    847         # The result is in the last temp var
--> 848         state.push(temps[-1])
    849 
    850     def op_BUILD_TUPLE_UNPACK_WITH_CALL(self, state, inst):

IndexError: Failed in nopython mode pipeline (step: analyzing bytecode)
list index out of range
```

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title, also removes some Python 2 and Python <3.6 restrictions.

Fixes #6534

### CODE REVIEW

File: numba/core/interpreter.py [hunk 1]
```diff
@@ -877,20 +877,23 @@ def op_CALL_FUNCTION_EX(self, inst, func, vararg, res):
...
-    def _build_tuple_unpack(self, inst, tuples, temps):
+    def _build_tuple_unpack(self, inst, tuples, temps, is_assign):
         first = self.get(tuples[0])
-        for other, tmp in zip(map(self.get, tuples[1:]), temps):
-            out = ir.Expr.binop(fn=operator.add, lhs=first, rhs=other,
-                                loc=self.loc)
-            self.store(out, tmp)
-            first = self.get(tmp)
+        if is_assign: # it's assign-like, just store through
+            self.store(first, temps[0])
```
Comment (sklam) [inline numba/core/interpreter.py:883]: There needs to be a convert to tuple.

```

In [5]: @njit
   ...: def foo(x):
   ...:     y = (*x,)
   ...:     return y
   ...:

In [6]: foo([1, 2])

Out[6]: [1, 2]
```

Comment (stuartarchibald) [reply numba/core/interpreter.py:883]: Is there actually enough information available to handle this?
```python
In [1]: import dis

In [2]: def foo(x):
   ...:     return (*x,)
   ...: 

In [3]: foo((1,2))
Out[3]: (1, 2)

In [4]: foo([1,2])
Out[4]: (1, 2)

In [5]: dis.dis(foo)
  2           0 LOAD_FAST                0 (x)
              2 BUILD_TUPLE_UNPACK       1
              4 RETURN_VALUE

```
the return type ends up depending on the list size.

Comment (sklam) [reply numba/core/interpreter.py:883]: Should we raise UnsupportedError/TypeError instead when the `x` in `(*x,)` is not a `tuple`?

Comment (stuartarchibald) [reply numba/core/interpreter.py:883]: I think it has to be an error if it's not a `tuple`. From OOB conversation, going try with emitting an intrinsic to handle this, can't do anything from the bytecode interpreter as there's nothing unique about the tuple vs. anything else situation.

Comment (stuartarchibald) [reply numba/core/interpreter.py]: Managed to get this to work in 28bdbca.

**Comment by sklam** [comment]
There are some py3.9 specific errors

**Review by sklam — APPROVED** [review]
Looks good. Thanks for the patch.

**Pull request merged**